src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,191 @@
+        if (tail != null) {
+            tail.next = newNode;
+        } else {
+            head = newNode;
+        }
+        tail = newNode;
+        size++;
+    }
+
+    private void addInsideList(int index, T value) {
+        Node<T> currentNode = getNodeByIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        addNodeInsideList(newNode, currentNode);
+    }
+
+    private void addNodeInsideList(Node<T> newNode, Node<T> oldNode) {
+        if (oldNode == head) {
+            addToFirstPosition(newNode.value);
+        } else {
+            newNode.prev = oldNode.prev;
+            oldNode.prev.next = newNode;
+            oldNode.prev = newNode;
+            newNode.next = oldNode;
+        }
+        size++;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("incorrect index");
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("incorrect index");
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        Node<T> currentNode = null;
+        if (index <= size / 2) {
+            int count = 0;
+            currentNode = head;
+            while (count != index) {
+                if (index == count++) {
+                    break;
+                }
+                currentNode = currentNode.next;
+            }
+        } else {
+            int count = size - 1;
+            currentNode = tail;
+            while (count != index) {
+                if (index == count--) {
+                    break;
+                }
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private boolean removeNode(Node<T> currentNode) {
+        if (currentNode == null) {
+            return false;
+        }
+        if (currentNode == head && currentNode == tail) {
+            head = null;
+            tail = null;
+        } else if (currentNode == head) {
+            if (currentNode.next != null) {
+                currentNode.next.prev = null;
+            }
+            head = currentNode.next;
+        } else if (currentNode == tail) {
+            currentNode.prev.next = null;
+            tail = currentNode.prev;
+        } else {
+            currentNode.next.prev = currentNode.prev;
+            currentNode.prev.next = currentNode.next;
+        }
+        size--;
+        return true;
+    }
+
+    private Node<T> getNodeByValue(T value) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (value != null
+                    && value.equals(currentNode.value) || value == currentNode.value) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+        }
+        return null;
+    }
+
+    private class Node<T> {
```

Comment:
```suggestion
    private static class Node<T> {
```

Reply:
if you can make an inner class static (there's no reference to enclosing class) - you should do it, better practice